src/commonMain/kotlin/org/kosat/NonIncremental.kt

```diff
@@ -0,0 +1,275 @@
+                        uselessClauses.add(it)
+                    }
+                }
+            }
+        }
+
+        clauses.removeAll(uselessClauses.map { clauses[it] })
+        countOccurrence()
+        updateSig()
+    }
+
+    // making restart to remove useless clauses
+    private fun makeRestart() {
+        numberOfRestarts++
+        restartNumber *= restartCoeff
+        level = 0
+        trail.clear()
+        units.clear()
+        watchers.forEach { it.clear() }
+        vars.forEachIndexed { ind, _ ->
+            delVariable(ind)
+        }
+
+        removeSubsumedClauses()
+        countOccurrence()
+
+        updateSig()
+
+        buildWatchers()
+    }
+
+    private val clauseLimit = 1000
+
+    private fun addResolvents(ind: Int) {
+        for (cl1 in litOccurrence[litPos(ind)]) {
+            for (cl2 in litOccurrence[litPos(-ind)]) {
+                val newClause = clauses[cl1].toMutableSet()
+                newClause.remove(ind)
+                // check if clause is tautology
+                if (clauses[cl2].any { newClause.contains(-it) }) {
+                    continue
+                }
+                newClause.addAll(clauses[cl2])
+                newClause.remove(-ind)
+                clauses.add(ArrayList(newClause))
+                for (lit in newClause) {
+                    litOccurrence[litPos(lit)].add(clauses.lastIndex)
+                }
+            }
+        }
+    }
+
+    private fun bve() {
+        val isLiteralRemoved = MutableList(varsNumber + 1) { false }
+        val newNumeration = MutableList(varsNumber + 1) { 0 }
+        var currentInd = 1
+        while (clauses.size < clauseLimit && currentInd <= varsNumber) {
+            if (litOccurrence[litPos(currentInd)].size * litOccurrence[litPos(-currentInd)].size <= clauseLimit) {
+                isLiteralRemoved[currentInd] = true
+                addResolvents(currentInd)
+            }
+            currentInd++
+        }
+        val deletedClauses = mutableListOf<Int>()
+        clauses.forEachIndexed { ind, clause ->
+            for (lit in clause) {
+                if (isLiteralRemoved[litIndex(lit)]) {
+                    deletedClauses.add(ind)
+                    break
+                }
+            }
+        }
+        clauses.removeAll(deletedClauses.map { clauses[it] })
+        var newSize = 0
+        for (ind in 1..varsNumber) {
+            if (!isLiteralRemoved[ind]) {
+                newSize++
+                newNumeration[ind] = newSize
+            }
+        }
+        for (clause in clauses) {
+            clause.forEachIndexed { ind, lit ->
+                if (lit > 0) {
+                    clause[ind] = newNumeration[lit]
+                } else {
+                    clause[ind] = -newNumeration[-lit]
+                }
+            }
+        }
+        varsNumber = newSize
+        countOccurrence()
+        updateSig()
+    }
+
+    private fun removeTautologies() {
```

Comment:
Docs?
What is a "tautology"? How are you "removing" it? When this method is safe to call, and when it is not?